cmd/kubeadm/app/cmd/phases/upgrade/apply/addons.go

```diff
@@ -0,0 +1,188 @@
+	return workflow.Phase{
+		Name:  "addon",
+		Short: "Install required addons for passing conformance tests",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Install all the addons",
+				InheritFlags:   getAddonPhaseFlags("all"),
+				RunAllSiblings: true,
+			},
+			{
+				Name:         "coredns",
+				Short:        "Install the CoreDNS addon to a Kubernetes cluster",
+				InheritFlags: getAddonPhaseFlags("coredns"),
+				Run:          runCoreDNSAddon,
+			},
+			{
+				Name:         "kube-proxy",
+				Short:        "Install the kube-proxy addon to a Kubernetes cluster",
+				InheritFlags: getAddonPhaseFlags("kube-proxy"),
+				Run:          runKubeProxyAddon,
+			},
+		},
+	}
+}
+
+func shouldUpgradeAddons(client clientset.Interface, cfg *kubeadmapi.InitConfiguration, out io.Writer) (bool, error) {
+	unupgradedControlPlanes, err := upgrade.UnupgradedControlPlaneInstances(client, cfg.NodeRegistration.Name)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to determine whether all the control plane instances have been upgraded")
+	}
+	if len(unupgradedControlPlanes) > 0 {
+		fmt.Fprintf(out, "[upgrade/addons] skip upgrade addons because control plane instances %v have not been upgraded\n", unupgradedControlPlanes)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func getInitData(c workflow.RunData) (*kubeadmapi.InitConfiguration, clientset.Interface, string, io.Writer, bool, error) {
+	data, ok := c.(Data)
+	if !ok {
+		return nil, nil, "", nil, false, errors.New("addon phase invoked with an invalid data struct")
+	}
+	return data.InitCfg(), data.Client(), data.PatchesDir(), data.OutputWriter(), data.DryRun(), nil
+}
+
+// runCoreDNSAddon installs CoreDNS addon to a Kubernetes cluster
+func runCoreDNSAddon(c workflow.RunData) error {
+	cfg, client, patchesDir, out, dryRun, err := getInitData(c)
+	if err != nil {
+		return err
+	}
+
+	shouldUpgradeAddons, err := shouldUpgradeAddons(client, cfg, out)
+	if err != nil {
+		return err
+	}
+	if !shouldUpgradeAddons {
+		return nil
+	}
+
+	// If the coredns ConfigMap is missing, show a warning and assume that the
+	// DNS addon was skipped during "kubeadm init", and that its redeployment on upgrade is not desired.
+	//
+	// TODO: remove this once "kubeadm upgrade apply" phases are supported:
+	//   https://github.com/kubernetes/kubeadm/issues/1318
+	if _, err := client.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(
+		context.TODO(),
+		kubeadmconstants.CoreDNSConfigMap,
+		metav1.GetOptions{},
+	); err != nil && apierrors.IsNotFound(err) {
+		klog.Warningf("the ConfigMaps %q in the namespace %q were not found. "+
+			"Assuming that a DNS server was not deployed for this cluster. "+
+			"Note that once 'kubeadm upgrade apply' supports phases you "+
+			"will have to skip the DNS upgrade manually",
+			kubeadmconstants.CoreDNSConfigMap,
+			metav1.NamespaceSystem)
+		return nil
+	}
```

Comment:
These lines have been removed now.

Reply:
We will need to update the document, and if the user needs to skip the addon installation during the upgrade phase, this will be done through `--skip-phases`.